contracts/wrapped-polyx/lib.rs

```diff
@@ -0,0 +1,456 @@
+            self.ensure_initialized()?;
+
+            // Get the caller's identity.
+            let caller_did = self.get_caller_did()?;
+
+            // Ensure the caller has a portfolio.
+            let caller_portfolio = self.ensure_has_portfolio(caller_did)?;
+
+            let api = Api::new();
+            // Get the next instruction id.
+            let instruction_id = api
+                .query()
+                .settlement()
+                .instruction_counter()
+                .map(|v| v.into())?;
+
+            // Transfer tokens to the caller's portfolio.
+            let our_portfolio = PortfolioId {
+              did: self.did,
+              kind: PortfolioKind::Default,
+            };
+            api.call().settlement().add_and_affirm_instruction(
+              self.venue,
+              SettlementType::SettleManual(0),
+              None,
+              None,
+              vec![Leg {
+                from: caller_portfolio,
+                to: our_portfolio,
+                asset: self.ticker,
+                amount: amount,
+              }],
+              vec![
+                our_portfolio,
+                caller_portfolio,
+              ],
+            ).submit()?;
+
+            api.call().settlement().execute_manual_instruction(
+                instruction_id,
+                1,
+                None
+            ).submit()?;
+
+            api.call().asset().redeem(
+              self.ticker,
+              amount,
+            ).submit()?;
+
+            if Self::env().transfer(Self::env().caller().into(), amount).is_err() {
+              panic!("error transferring")
+            }
+
+            Self::env().emit_event(PolyxUnwrapped {
+              did: caller_did,
+              key: Self::env().caller().into(),
+              amount: amount,
+            });
+
+            Ok(())
+        }
+
+        #[ink(message)]
+        /// Return the caller's portfolio custodianship back to them.
+        pub fn remove_portfolio(&mut self) -> Result<()> {
+            self.ensure_initialized()?;
+
+            // Get the caller's identity.
+            let caller_did = self.get_caller_did()?;
+
+            // Ensure the caller has a portfolio.
+            let portfolio = self.ensure_has_portfolio(caller_did)?;
+
+            let api = Api::new();
+            // Remove our custodianship.
+            api.call().portfolio().quit_portfolio_custody(portfolio).submit()?;
+            // Remove the portfolio.
+            self.portfolios.remove(caller_did);
+            Self::env().emit_event(PortfolioRemoved {
+              did: caller_did,
+              portfolio_kind: portfolio.kind,
+            });
+
+            Ok(())
+        }
+
+        #[ink(message)]
+        /// Allow the caller to withdrawal funds from the contract controlled portfolio.
+        pub fn withdraw_polyx(
+            &mut self,
+            amount: Balance,
+            dest: PortfolioKind,
+        ) -> Result<()> {
```

Comment:
This could be merged with the `mint_wrapped_polyx` call above, since most users would need to do both of these each time they want to wrap POLYX.

Reply:
Just make sure that `dest` portfolio is not the same as the custodian portfolio.  This is mostly to save on gas fees, since the last step to move funds would fail and revert the whole transaction.